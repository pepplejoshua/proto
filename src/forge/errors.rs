#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
pub enum ForgeError {}
