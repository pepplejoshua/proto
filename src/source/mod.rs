pub mod errors;
pub mod source;
