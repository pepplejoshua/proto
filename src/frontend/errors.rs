use super::source::SourceRef;

#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
pub enum LexError {
    InvalidCharacter(SourceRef),
    CannotMakeSignedNumber(SourceRef),
    CannotMakeUnsignedNumber(SourceRef),
}

#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
pub enum ParseError {
    Expected(String, SourceRef, Option<String>),
    ConstantDeclarationNeedsInitValue(SourceRef),
    CannotParseAnExpression(SourceRef),
    TooManyFnArgs(SourceRef),
    MalformedDeclaration(String, SourceRef),
}
