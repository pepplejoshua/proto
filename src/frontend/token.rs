use super::source::SourceRef;

#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug, Clone)]
pub enum Token {
    // keywords
    Fn(SourceRef),
    Let(SourceRef),
    Mut(SourceRef),
    If(SourceRef),
    Else(SourceRef),
    Loop(SourceRef),
    While(SourceRef),
    Break(SourceRef),
    Continue(SourceRef),
    True(SourceRef),
    False(SourceRef),
    Pub(SourceRef),
    Return(SourceRef),
    Comp(SourceRef),

    // operators
    Plus(SourceRef),
    Minus(SourceRef),
    Star(SourceRef),
    Slash(SourceRef),
    Modulo(SourceRef),

    // special characters
    At(SourceRef),

    // comparison
    Equal(SourceRef),
    NotEqual(SourceRef),
    Less(SourceRef),
    LessEqual(SourceRef),
    Greater(SourceRef),
    GreaterEqual(SourceRef),

    // assignment
    Assign(SourceRef),

    // logical
    And(SourceRef),
    Or(SourceRef),
    Not(SourceRef),

    // delimiters
    LParen(SourceRef),
    RParen(SourceRef),
    LCurly(SourceRef),
    RCurly(SourceRef),
    LBracket(SourceRef),
    RBracket(SourceRef),
    Semicolon(SourceRef),
    Colon(SourceRef),
    Comma(SourceRef),
    Dot(SourceRef),

    // literals
    NumberLiteral(String, SourceRef),
    Identifier(String, SourceRef),
    CharLiteral(SourceRef, char),
    StringLiteral(SourceRef, String),
    SingleLineStringLiteral(SourceRef, String),
    MultiLineStringFragment(SourceRef, String),
    SingleLineComment(SourceRef, String),

    // type tags
    I8(SourceRef),
    I16(SourceRef),
    I32(SourceRef),
    I64(SourceRef),
    ISize(SourceRef),
    U8(SourceRef),
    U16(SourceRef),
    U32(SourceRef),
    U64(SourceRef),
    USize(SourceRef),
    Bool(SourceRef),
    Void(SourceRef),
    Char(SourceRef),
    Str(SourceRef),
    Type(SourceRef),
    Struct(SourceRef),

    // misc
    Eof(SourceRef),
}

#[allow(dead_code)]
impl Token {
    pub fn get_source_ref(&self) -> SourceRef {
        match self {
            Token::SingleLineComment(src, _) => src.clone(),
            Token::Fn(src) => src.clone(),
            Token::Let(src) => src.clone(),
            Token::Mut(src) => src.clone(),
            Token::If(src) => src.clone(),
            Token::Else(src) => src.clone(),
            Token::Loop(src) => src.clone(),
            Token::While(src) => src.clone(),
            Token::Plus(src) => src.clone(),
            Token::Minus(src) => src.clone(),
            Token::Star(src) => src.clone(),
            Token::Slash(src) => src.clone(),
            Token::Modulo(src) => src.clone(),
            Token::Equal(src) => src.clone(),
            Token::NotEqual(src) => src.clone(),
            Token::Less(src) => src.clone(),
            Token::LessEqual(src) => src.clone(),
            Token::Greater(src) => src.clone(),
            Token::GreaterEqual(src) => src.clone(),
            Token::Assign(src) => src.clone(),
            Token::And(src) => src.clone(),
            Token::Or(src) => src.clone(),
            Token::Not(src) => src.clone(),
            Token::LParen(src) => src.clone(),
            Token::RParen(src) => src.clone(),
            Token::LCurly(src) => src.clone(),
            Token::RCurly(src) => src.clone(),
            Token::LBracket(src) => src.clone(),
            Token::RBracket(src) => src.clone(),
            Token::Semicolon(src) => src.clone(),
            Token::Comma(src) => src.clone(),
            Token::Eof(src) => src.clone(),
            Token::Void(src) => src.clone(),
            Token::True(src) => src.clone(),
            Token::False(src) => src.clone(),
            Token::CharLiteral(src, _) => src.clone(),
            Token::Identifier(_, src) => src.clone(),
            Token::Dot(src) => src.clone(),
            Token::Break(src) => src.clone(),
            Token::Continue(src) => src.clone(),
            Token::Pub(src) => src.clone(),
            Token::Return(src) => src.clone(),
            Token::At(src) => src.clone(),
            Token::SingleLineStringLiteral(src, _) => src.clone(),
            Token::MultiLineStringFragment(src, _) => src.clone(),
            Token::NumberLiteral(_, src) => src.clone(),
            Token::Comp(src) => src.clone(),
            Token::Struct(src) => src.clone(),
            Token::I8(src) => src.clone(),
            Token::I16(src) => src.clone(),
            Token::I32(src) => src.clone(),
            Token::I64(src) => src.clone(),
            Token::ISize(src) => src.clone(),
            Token::U8(src) => src.clone(),
            Token::U16(src) => src.clone(),
            Token::U32(src) => src.clone(),
            Token::U64(src) => src.clone(),
            Token::USize(src) => src.clone(),
            Token::Bool(src) => src.clone(),
            Token::Char(src) => src.clone(),
            Token::Str(src) => src.clone(),
            Token::StringLiteral(src, _) => src.clone(),
            Token::Type(src) => src.clone(),
            Token::Colon(src) => src.clone(),
        }
    }

    pub fn is_terminator(&self) -> bool {
        matches!(self, Token::Semicolon(_) | Token::RCurly(_) | Token::Eof(_))
    }

    pub fn begins_instruction(&self) -> bool {
        matches!(
            self,
            Token::Let(_)
                | Token::LCurly(_)
                | Token::Loop(_)
                | Token::While(_)
                | Token::Pub(_)
                | Token::Return(_)
                | Token::Break(_)
                | Token::Continue(_)
                | Token::If(_)
                | Token::Struct(_)
                | Token::At(_)
                | Token::SingleLineComment(_, _)
                | Token::Fn(_)
        )
    }

    pub fn as_str(&self) -> String {
        match self {
            Token::Fn(_) => "fn".into(),
            Token::Let(_) => "let".into(),
            Token::Mut(_) => "mut".into(),
            Token::If(_) => "if".into(),
            Token::Else(_) => "else".into(),
            Token::Loop(_) => "loop".into(),
            Token::While(_) => "while".into(),
            Token::Break(_) => "break".into(),
            Token::Continue(_) => "continue".into(),
            Token::Void(_) => "void".into(),
            Token::True(_) => "true".into(),
            Token::False(_) => "false".into(),
            Token::CharLiteral(_, c) => format!("'{c}'"),
            Token::Plus(_) => "+".into(),
            Token::Minus(_) => "-".into(),
            Token::Star(_) => "*".into(),
            Token::Slash(_) => "/".into(),
            Token::Modulo(_) => "%".into(),
            Token::Equal(_) => "==".into(),
            Token::NotEqual(_) => "!=".into(),
            Token::Less(_) => "<".into(),
            Token::LessEqual(_) => "<=".into(),
            Token::Greater(_) => ">".into(),
            Token::GreaterEqual(_) => ">=".into(),
            Token::Assign(_) => "=".into(),
            Token::And(_) => "and".into(),
            Token::Or(_) => "or".into(),
            Token::Not(_) => "!".into(),
            Token::LParen(_) => "(".into(),
            Token::RParen(_) => ")".into(),
            Token::LCurly(_) => "{".into(),
            Token::RCurly(_) => "}".into(),
            Token::LBracket(_) => "[".into(),
            Token::RBracket(_) => "]".into(),
            Token::Semicolon(_) => ";".into(),
            Token::Comma(_) => ",".into(),
            Token::Dot(_) => ".".into(),
            Token::Identifier(name, _) => name.clone(),
            Token::NumberLiteral(num, _) => num.to_string(),
            Token::I8(_) => "i8".into(),
            Token::I16(_) => "i16".into(),
            Token::I32(_) => "i32".into(),
            Token::I64(_) => "i64".into(),
            Token::ISize(_) => "isize".into(),
            Token::U8(_) => "u8".into(),
            Token::U16(_) => "u16".into(),
            Token::U32(_) => "u32".into(),
            Token::U64(_) => "u64".into(),
            Token::USize(_) => "usize".into(),
            Token::Bool(_) => "bool".into(),
            Token::Char(_) => "char".into(),
            Token::Str(_) => "str".into(),
            Token::Eof(_) => "\0".into(),
            Token::Pub(_) => "pub".into(),
            Token::Return(_) => "return".into(),
            Token::At(_) => "@".into(),
            Token::StringLiteral(_, src) => format!("\"{src}\""),
            Token::SingleLineComment(_, src) => src.clone(),
            Token::Type(_) => "type".into(),
            Token::Struct(_) => "struct".into(),
            Token::Comp(_) => "comp".into(),
            Token::Colon(_) => ":".into(),
            Token::SingleLineStringLiteral(_, content) => content.clone(),
            Token::MultiLineStringFragment(_, mls) => mls.clone(),
        }
    }
}
