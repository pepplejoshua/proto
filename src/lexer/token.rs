use std::rc::Rc;

use crate::source::source::{SourceFile, SourceRef};

#[allow(dead_code, unused)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TokenType {
    // keywords
    Pub,
    Fn,
    Var,
    If,
    Else,
    For,
    In,
    Break,
    Continue,
    True,
    False,
    Return,
    Struct,
    Print,
    Println,
    Some,
    None,
    Defer,
    Const,
    Comptime,
    Use,
    Module,
    As,
    Root,

    // operators
    BackSlash,
    Plus,
    Minus,
    Star,
    Slash,
    Modulo,
    PlusAssign,
    MinusAssign,
    StarAssign,
    SlashAssign,
    ModuloAssign,
    Ampersand,
    Equal,
    NotEqual,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Assign,
    And,
    Or,
    Not,

    // special characters
    At,

    // delimiters
    LParen,
    RParen,
    LCurly,
    RCurly,
    LBracket,
    RBracket,
    Semicolon,
    Colon,
    DoubleColon,
    ColonAssign,
    Comma,
    Dot,
    BackTick,

    // complex tokens
    Integer,
    Decimal,
    Identifier,
    Character,
    String,
    InterpolatedStringFragment,
    Comment,

    // type tags
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    F32,
    F64,
    Bool,
    Void,
    Char,
    Str,
    Type,

    // misc
    Underscore,
    QuestionMark,
    Eof,
}

#[allow(unused)]
#[derive(Debug, Clone)]
pub struct SrcToken {
    pub ty: TokenType,
    pub loc: Rc<SourceRef>,
}

#[allow(unused)]
impl SrcToken {
    pub fn get_source_ref(self: &SrcToken) -> Rc<SourceRef> {
        return self.loc.clone();
    }

    pub fn as_str(self: &SrcToken, src: &SourceFile) -> String {
        match self.ty {
            TokenType::As => "as".to_string(),
            TokenType::Module => "module".to_string(),
            TokenType::Use => "use".to_string(),
            TokenType::Pub => "pub".to_string(),
            TokenType::Comptime => "comptime".to_string(),
            TokenType::Fn => "fn".to_string(),
            TokenType::Const => "const".to_string(),
            TokenType::Var => "var".to_string(),
            TokenType::If => "if".to_string(),
            TokenType::Else => "else".to_string(),
            TokenType::For => "for".to_string(),
            TokenType::In => "in".to_string(),
            TokenType::Break => "break".to_string(),
            TokenType::Continue => "continue".to_string(),
            TokenType::True => "true".to_string(),
            TokenType::False => "false".to_string(),
            TokenType::Return => "return".to_string(),
            TokenType::Struct => "struct".to_string(),
            TokenType::Print => "print".to_string(),
            TokenType::Println => "println".to_string(),
            TokenType::Some => "some".to_string(),
            TokenType::None => "none".to_string(),
            TokenType::Defer => "defer".to_string(),
            TokenType::Const => "const".to_string(),
            TokenType::BackSlash => "\\".to_string(),
            TokenType::Plus => "+".to_string(),
            TokenType::Minus => "-".to_string(),
            TokenType::Star => "*".to_string(),
            TokenType::Slash => "/".to_string(),
            TokenType::Modulo => "%".to_string(),
            TokenType::PlusAssign => "+=".to_string(),
            TokenType::MinusAssign => "-=".to_string(),
            TokenType::StarAssign => "*=".to_string(),
            TokenType::SlashAssign => "/=".to_string(),
            TokenType::ModuloAssign => "%=".to_string(),
            TokenType::Ampersand => "&".to_string(),
            TokenType::Equal => "==".to_string(),
            TokenType::NotEqual => "!=".to_string(),
            TokenType::Less => "<".to_string(),
            TokenType::LessEqual => "<=".to_string(),
            TokenType::Greater => ">".to_string(),
            TokenType::GreaterEqual => ">=".to_string(),
            TokenType::Assign => "=".to_string(),
            TokenType::And => "and".to_string(),
            TokenType::Or => "or".to_string(),
            TokenType::Not => "!".to_string(),
            TokenType::At => "@".to_string(),
            TokenType::LParen => "(".to_string(),
            TokenType::RParen => ")".to_string(),
            TokenType::LCurly => "{".to_string(),
            TokenType::RCurly => "}".to_string(),
            TokenType::LBracket => "[".to_string(),
            TokenType::RBracket => "]".to_string(),
            TokenType::Semicolon => ";".to_string(),
            TokenType::Colon => ":".to_string(),
            TokenType::DoubleColon => "::".to_string(),
            TokenType::ColonAssign => ":=".to_string(),
            TokenType::Comma => ",".to_string(),
            TokenType::Dot => ".".to_string(),
            TokenType::BackTick => "`".to_string(),
            TokenType::Integer
            | TokenType::Decimal
            | TokenType::Identifier
            | TokenType::InterpolatedStringFragment
            | TokenType::Comment
            | TokenType::String
            | TokenType::Character => {
                let snippet = &src.text[self.loc.flat_start..self.loc.flat_end];
                snippet.to_string()
            }
            TokenType::I8 => "i8".to_string(),
            TokenType::I16 => "i16".to_string(),
            TokenType::I32 => "i32".to_string(),
            TokenType::I64 => "i64".to_string(),
            TokenType::U8 => "u8".to_string(),
            TokenType::U16 => "u16".to_string(),
            TokenType::U32 => "u32".to_string(),
            TokenType::U64 => "u64".to_string(),
            TokenType::F32 => "f32".to_string(),
            TokenType::F64 => "f64".to_string(),
            TokenType::Bool => "bool".to_string(),
            TokenType::Void => "void".to_string(),
            TokenType::Char => "char".to_string(),
            TokenType::Str => "str".to_string(),
            TokenType::Type => "type".to_string(),
            TokenType::Underscore => "_".to_string(),
            TokenType::QuestionMark => "?".to_string(),
            TokenType::Eof => "\0".to_string(),
        }
    }
}
