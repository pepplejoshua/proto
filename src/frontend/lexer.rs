use super::errors::LexError;
use super::source::SourceFile;
use super::token::Token;

#[allow(dead_code)]
pub struct Lexer {
    pub src: SourceFile,
    last_lexed: Option<Token>,
}

#[allow(dead_code)]
impl Lexer {
    pub fn new(src: SourceFile) -> Lexer {
        Lexer {
            src,
            last_lexed: None,
        }
    }

    // skip all whitespace characters
    fn skip_whitespace(&mut self) {
        loop {
            let c = self.src.cur_char();
            if c.is_whitespace() {
                self.src.next_char();
                continue;
            } else {
                break;
            }
        }
    }

    // process comments
    fn process_comment(&mut self) {
        let start = self.src.cur_char();
        let peek = self.src.peek_char();

        if start == '/' && peek == '/' {
            // consume both "//" characters
            self.src.next_char();
            self.src.next_char();

            // single line comment
            // skip until we reach a newline
            loop {
                let cur = self.src.cur_char();
                self.src.next_char();
                if cur != '\n' {
                    continue;
                } else {
                    break;
                }
            }
        }
    }

    // try to lex the next possible token
    pub fn next_token(&mut self) -> Result<Token, LexError> {
        while self.src.cur_char().is_whitespace()
            || (self.src.cur_char() == '/' && self.src.peek_char() == '/')
        {
            // skip all preceding whitespace
            self.skip_whitespace();
            // process comments
            // eventually, it'll get documentation comments
            // for now, it just skips single line comments
            self.process_comment();
        }

        // get the current character
        let c = self.src.cur_char();

        // if we reached the end of the file, return EOF
        if c == '\0' {
            return Ok(Token::Eof(self.src.get_ref()));
        }
        let maybe_token = match c {
            // operators
            '+' | '-' | '*' | '/' | '%' | '!' | '=' | '<' | '>' | '(' | ')' | '{' | '}' | '['
            | ']' | ',' | '.' | ':' | ';' | '^' | '$' | '@' => self.lex_operator(),
            // numbers
            _ if c.is_ascii_digit() => self.lex_number(),
            // identifiers | keywords
            _ if c.is_alphabetic() || c == '_' => self.lex_potential_identifier(),
            // ' which is used to start characters or strings
            '\'' => self.lex_char(),
            // " which is used to start strings
            '"' => self.lex_string(),
            // invalid character
            _ => {
                let mut err_ref = self.src.get_ref();
                err_ref.end_col += 1;
                let err = Err(LexError::InvalidCharacter(err_ref));
                self.src.next_char(); // try to keep lexing
                err
            }
        };

        match &maybe_token {
            Ok(t) => self.last_lexed = Some(t.clone()),
            Err(_) => self.last_lexed = None,
        }
        maybe_token
    }

    // lex a potential identifier
    // it might turn out to be either a:
    // - keyword
    // - identifier
    fn lex_potential_identifier(&mut self) -> Result<Token, LexError> {
        let mut id = String::new();
        let c_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        id.push(cur);

        // read all characters that are alphanumeric or '_'
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_alphanumeric() || c == '_' {
                id.push(c);
            } else {
                break;
            }
        }

        let combined_ref = c_ref.combine(end_ref);
        // check if the identifier is a keyword
        match id.as_str() {
            "fn" => Ok(Token::Fn(combined_ref)),
            "let" => Ok(Token::Let(combined_ref)),
            "mut" => Ok(Token::Mut(combined_ref)),
            "if" => Ok(Token::If(combined_ref)),
            "else" => Ok(Token::Else(combined_ref)),
            "loop" => Ok(Token::Loop(combined_ref)),
            "while" => Ok(Token::While(combined_ref)),
            "void" => Ok(Token::Void(combined_ref)),
            "true" => Ok(Token::True(combined_ref)),
            "false" => Ok(Token::False(combined_ref)),
            "break" => Ok(Token::Break(combined_ref)),
            "continue" => Ok(Token::Continue(combined_ref)),
            "return" => Ok(Token::Return(combined_ref)),
            "use" => Ok(Token::Use(combined_ref)),
            "pub" => Ok(Token::Pub(combined_ref)),
            "mod" => Ok(Token::Mod(combined_ref)),
            "as" => Ok(Token::As(combined_ref)),

            "and" => Ok(Token::And(combined_ref)),
            "or" => Ok(Token::Or(combined_ref)),
            "not" => Ok(Token::Not(combined_ref)),

            "i8" => Ok(Token::I8(combined_ref)),
            "i16" => Ok(Token::I16(combined_ref)),
            "i32" => Ok(Token::I32(combined_ref)),
            "i64" => Ok(Token::I64(combined_ref)),
            "isize" => Ok(Token::Isize(combined_ref)),
            "u8" => Ok(Token::U8(combined_ref)),
            "u16" => Ok(Token::U16(combined_ref)),
            "u32" => Ok(Token::U32(combined_ref)),
            "u64" => Ok(Token::U64(combined_ref)),
            "usize" => Ok(Token::Usize(combined_ref)),
            "bool" => Ok(Token::Bool(combined_ref)),
            "char" => Ok(Token::Char(combined_ref)),
            "str" => Ok(Token::Str(combined_ref)),
            _ => Ok(Token::Identifier(id, combined_ref)),
        }
    }

    // lex an operator
    fn lex_operator(&mut self) -> Result<Token, LexError> {
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        // detect a signed number
        if cur == '-' {
            // if the next character is a digit, lex a signed number
            let peek = self.src.peek_char();
            if peek.is_ascii_digit() {
                return self.lex_signed_number();
            } else {
                // otherwise, return a minus operator
                self.src.next_char();
                return Ok(Token::Minus(cur_ref.combine(self.src.get_ref())));
            }
        }

        match cur {
            // arithmetic operators
            '+' => {
                self.src.next_char();
                Ok(Token::Plus(cur_ref.combine(self.src.get_ref())))
            }
            '-' => {
                self.src.next_char();
                Ok(Token::Minus(cur_ref.combine(self.src.get_ref())))
            }
            '*' => {
                self.src.next_char();
                Ok(Token::Star(cur_ref.combine(self.src.get_ref())))
            }
            '/' => {
                self.src.next_char();
                Ok(Token::Slash(cur_ref.combine(self.src.get_ref())))
            }
            '%' => {
                self.src.next_char();
                Ok(Token::Modulo(cur_ref.combine(self.src.get_ref())))
            }

            // special characters
            '^' => {
                self.src.next_char();
                Ok(Token::Caret(cur_ref.combine(self.src.get_ref())))
            }
            '@' => {
                self.src.next_char();
                Ok(Token::At(cur_ref.combine(self.src.get_ref())))
            }
            '$' => {
                self.src.next_char();
                Ok(Token::Dollar(cur_ref.combine(self.src.get_ref())))
            }

            // logical operators
            '!' => {
                // if the next character is a '=', return a NotEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::NotEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Not operator
                self.src.next_char();
                Ok(Token::Exclamation(cur_ref.combine(self.src.get_ref())))
            }
            '=' => {
                // if the next character is a '=', return a Equal operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Equal(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Assign operator
                self.src.next_char();
                Ok(Token::Assign(cur_ref.combine(self.src.get_ref())))
            }
            '<' => {
                // if the next character is a '=', return a LessEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::LessEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Less operator
                self.src.next_char();
                Ok(Token::Less(cur_ref.combine(self.src.get_ref())))
            }
            '>' => {
                // if the next character is a '=', return a GreaterEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::GreaterEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Greater operator
                self.src.next_char();
                Ok(Token::Greater(cur_ref.combine(self.src.get_ref())))
            }
            ':' => {
                // if the next character is a ':', return a Scope operator
                let c = self.src.peek_char();
                if c == ':' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Scope(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Colon operator
                self.src.next_char();
                Ok(Token::Colon(cur_ref.combine(self.src.get_ref())))
            }
            '(' => {
                self.src.next_char();
                Ok(Token::LParen(cur_ref.combine(self.src.get_ref())))
            }
            ')' => {
                self.src.next_char();
                Ok(Token::RParen(cur_ref.combine(self.src.get_ref())))
            }
            '{' => {
                self.src.next_char();
                Ok(Token::LCurly(cur_ref.combine(self.src.get_ref())))
            }
            '}' => {
                self.src.next_char();
                Ok(Token::RCurly(cur_ref.combine(self.src.get_ref())))
            }
            '[' => {
                self.src.next_char();
                Ok(Token::LBracket(cur_ref.combine(self.src.get_ref())))
            }
            ']' => {
                self.src.next_char();
                Ok(Token::RBracket(cur_ref.combine(self.src.get_ref())))
            }
            ',' => {
                self.src.next_char();
                Ok(Token::Comma(cur_ref.combine(self.src.get_ref())))
            }
            '.' => {
                self.src.next_char();
                Ok(Token::Dot(cur_ref.combine(self.src.get_ref())))
            }
            ';' => {
                self.src.next_char();
                Ok(Token::Semicolon(cur_ref.combine(self.src.get_ref())))
            }
            _ => unreachable!("invalid operator character: '{}' at {:?}", cur, cur_ref),
        }
    }

    fn lex_string(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read all characters until terminating "
        while !self.src.is_eof() {
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '"' {
                break;
            } else if c == '\\' {
                // if the next character is a ", add it to the content
                let c = self.src.next_char();
                span = span.combine(self.src.get_ref());
                if c == '"' {
                    content.push('"');
                } else {
                    // otherwise, add the escape character
                    content.push('\\');
                    content.push(c);
                }
            } else {
                content.push(c);
            }
        }

        // if the string is not terminated, return an error
        if self.src.is_eof() {
            Err(LexError::UnterminatedStringLiteral(span))
        } else {
            self.src.next_char();
            span = span.combine(self.src.get_ref());
            Ok(Token::StringLiteral(span, content))
        }
    }

    // try to lex a character
    fn lex_char(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read single character
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());

        // if the next character is a ', return an error
        if c == '\'' {
            return Err(LexError::EmptyCharacterLiteral(span));
        } else if c == '\\' {
            // if the next character is a ', add it to the content
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '\'' {
                content.push('\'');
            } else {
                // otherwise, add the escape character
                content.push('\\');
                content.push(c);
            }
        } else {
            content.push(c);
        }

        // if the next character is not a ', return an error
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());
        if c != '\'' {
            return Err(LexError::UnterminatedCharacterLiteral(span));
        }

        self.src.next_char();
        span = span.combine(self.src.get_ref());
        Ok(Token::CharLiteral(span, content.chars().next().unwrap()))
    }

    // try to lex a number
    fn lex_number(&mut self) -> Result<Token, LexError> {
        // by default, calling lex_number will lex a signed number
        // since users tend to perform signed arithmetic more often
        self.lex_signed_number()
    }

    // try to lex an unsigned number which is one of:
    // - i8
    // - i16
    // - i32
    // - i64
    // - isize
    fn lex_signed_number(&mut self) -> Result<Token, LexError> {
        let mut signed_number = String::new();
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        signed_number.push(cur);

        // read all characters that are digits
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_ascii_digit() {
                signed_number.push(c);
            } else if c == '_' {
                // ignore underscores
                continue;
            } else {
                break;
            }
        }

        let combined_ref = cur_ref.combine(end_ref);
        // check if signed_number is i8
        if let Ok(num) = signed_number.parse::<i8>() {
            return Ok(Token::I8Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i16>() {
            return Ok(Token::I16Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i32>() {
            return Ok(Token::I32Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i64>() {
            return Ok(Token::I64Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<isize>() {
            return Ok(Token::IsizeLiteral(num, combined_ref));
        }

        // if signed_number's first character is not '-',
        // then attempt to lex an unsigned number
        self.src.jump_to(&cur_ref);
        self.lex_unsigned_number()
    }

    // try to lex an unsigned number which is one of:
    // - u8
    // - u16
    // - u32
    // - u64
    // - usize
    fn lex_unsigned_number(&mut self) -> Result<Token, LexError> {
        let mut unsigned_number = String::new();
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        unsigned_number.push(cur);

        // read all characters that are digits
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_ascii_digit() {
                unsigned_number.push(c);
            } else if c == '_' {
                // ignore underscores
                continue;
            } else {
                break;
            }
        }

        let combined_ref = cur_ref.combine(end_ref);
        // check if unsigned_number is u8
        if let Ok(num) = unsigned_number.parse::<u8>() {
            return Ok(Token::U8Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u16>() {
            return Ok(Token::U16Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u32>() {
            return Ok(Token::U32Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u64>() {
            return Ok(Token::U64Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<usize>() {
            return Ok(Token::UsizeLiteral(num, combined_ref));
        }

        // if it is not a valid unsigned number, return an error
        Err(LexError::CannotMakeUnsignedNumber(combined_ref))
    }
}

#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug)]
struct LexerTestResult {
    stringified_tokens: Vec<String>,
}

#[test]
fn test_lexer() {
    insta::glob!("lexer_inputs/*.pr", |path| {
        // build the SourceFile from the proto file
        let path = path.to_str().unwrap().to_string();
        let src: SourceFile = SourceFile::new(path);

        // build the lexer
        let mut lexer = Lexer::new(src);

        let mut tokens = Vec::new();
        let mut res = LexerTestResult {
            stringified_tokens: Vec::new(),
        };
        loop {
            let token = lexer.next_token();
            if token.is_err() {
                panic!("error lexing: {token:?}");
            }
            let token = token.unwrap();
            if let Token::Eof(_) = token {
                break;
            }
            res.stringified_tokens.push(format!(
                "{} at {}",
                token.as_str(),
                token.get_source_ref().as_str()
            ));
            tokens.push(token);
        }
        insta::assert_yaml_snapshot!(res)
    });
}
