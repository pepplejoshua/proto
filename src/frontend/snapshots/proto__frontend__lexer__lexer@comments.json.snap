---
source: src/frontend/lexer.rs
expression: res
input_file: src/frontend/lexer_inputs/comments.json
---
stringified_tokens:
  - "Identifier(\"some_ident\", SourceRef { file: \"comments.json.pr\", start_line: 1, start_col: 0, end_line: 1, end_col: 10, flat_start: 21, flat_end: 31 })"
  - "False(SourceRef { file: \"comments.json.pr\", start_line: 1, start_col: 11, end_line: 1, end_col: 16, flat_start: 32, flat_end: 37 })"
  - "Identifier(\"some_other_ident\", SourceRef { file: \"comments.json.pr\", start_line: 3, start_col: 0, end_line: 3, end_col: 16, flat_start: 65, flat_end: 81 })"
  - "True(SourceRef { file: \"comments.json.pr\", start_line: 3, start_col: 17, end_line: 3, end_col: 21, flat_start: 82, flat_end: 86 })"

